### PR TITLE
[0.61] Enable integer-type native runtime settings.

### PR DESCRIPTION
**Pull request by JunielKatarn**
Backport of #5669

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5713)

### CODE REVIEW

File: vnext/Desktop.DLL/react-native-win32.x64.def [hunk 1]
```diff
@@ -50,6 +50,16 @@ EXPORTS
 ?typeName@dynamic@folly@@QEBAPEBDXZ
 ?updateProperties@ShadowNode@react@facebook@@UEAAX$$QEBUdynamic@folly@@@Z
 ?Utf16ToUtf8@Unicode@Common@Microsoft@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@5@@Z
+?SetRuntimeOptionBool@React@Microsoft@@YAX$$QEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_N@Z
+?SetRuntimeOptionInt@React@Microsoft@@YAX$$QEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@H@Z
```
Comment (aeulitz) [inline vnext/Desktop.DLL/react-native-win32.x64.def:54]: Is this specific to a particular native module? If it's part of the general "instance management" API, will we have to transcribe this into the C++/WinRT ABI?

File: vnext/Desktop.UnitTests/UtilsTest.cpp [hunk 2]
```diff
@@ -25,7 +24,7 @@ TEST_CLASS(UtilsTest) {
       string path = "/",
       string query = "") {
 
-    Url url(urlString);
+    Url url(std::move(urlString));
```
Comment (aeulitz) [inline vnext/Desktop.UnitTests/UtilsTest.cpp:27]: >std::move(urlString) [](start = 12, length = 20)

Can we take this one step further and declare the 'urlString' parameter as an r-value reference

File: vnext/Shared/RuntimeOptions.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/// <summary>
+/// Sets a global boolean value identified by an arbitrary string.
+/// </summary>
+/// <param name="name">Global key</param>
+void __cdecl SetRuntimeOptionBool(std::string &&name, bool value) noexcept;
+
+/// <summary>
+/// Sets a global signed integer value identified by an arbitrary string.
+/// </summary>
+/// <param name="name">Global boolean key</param>
```
Comment (aeulitz) [inline vnext/Shared/RuntimeOptions.h:19]: ><param name="name">Global boolean key</param> [](start = 4, length = 45)

Nit: doesn't seem to line up with actual parameters.

File: vnext/Shared/RuntimeOptions.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+void __cdecl SetRuntimeOptionInt(std::string &&name, std::int32_t value) noexcept;
+
+/// <summary>
+/// Retrieves a global boolean value for the given key.
+/// </summary>
+/// <param name="name">Global boolean key</param>
+/// <returns>Value stored for the given key, or false if the entry doesn't exist (default)</returns>
+const bool __cdecl GetRuntimeOptionBool(const std::string &name) noexcept;
```
Comment (aeulitz) [inline vnext/Shared/RuntimeOptions.h:27]: >const bool [](start = 0, length = 10)

For my education: What's the benefit of declaring a value return type 'const'?

File: vnext/Shared/RuntimeOptions.h (hunk 3 above)
Comment (JunielKatarn) [reply vnext/Shared/RuntimeOptions.h:19]: True. I'll have to update it in master too.

File: vnext/Shared/RuntimeOptions.h (hunk 4 above)
Comment (aeulitz) [inline vnext/Shared/RuntimeOptions.h:27]: >const std::string &name [](start = 40, length = 23)

Is a std::string needed/advantageous in the function implementation? If not, wouldn't a 'const char*' suffice, and potentially prevent a std::string allocation at the call sites?

Comment (JunielKatarn) [reply vnext/Shared/RuntimeOptions.h:27]: Good point. This API would likely just use a string literal as argument, and not a variable.
We may change this to `const char*`, or `std::string_view`.

File: vnext/Desktop.UnitTests/UtilsTest.cpp (hunk 2 above)
Comment (JunielKatarn) [reply vnext/Desktop.UnitTests/UtilsTest.cpp:27]: Sounds good, but being a backport, I prefer to change this in a subsequent, master-targeting PR.

File: vnext/Shared/RuntimeOptions.h (hunk 4 above)
Comment (JunielKatarn) [reply vnext/Shared/RuntimeOptions.h:27]: Can't remember the details, but @NikoAri suggested this.
I think it was related to the value guaranteed to not be modified in the invoked function allowing for a compile-time optimization.

File: vnext/Desktop.DLL/react-native-win32.x64.def (hunk 1 above)
Comment (JunielKatarn) [reply vnext/Desktop.DLL/react-native-win32.x64.def:54]: It's a generic mechanism to configure arbitrary settings on runtime in any part of the native code base.
According to @acoates-ms, this should be available "out of the box" in RN 0.63, so these functions are transitory.

File: vnext/Shared/RuntimeOptions.h (hunk 4 above)
Comment (NikoAri) [reply vnext/Shared/RuntimeOptions.h:27]: std::string_view would be nice!

Returning "const bool" shouldn't be needed, "bool" should be good!

---
In reply to: [470318324](https://github.com/microsoft/react-native-windows/pull/5713#discussion_r470318324) [](ancestors = 470318324)

**Pull request merged**